File: backend/src/clientbridge/models/billing.py
from datetime import date, datetime

from sqlalchemy import (
    BigInteger,
    Date,
    DateTime,
    ForeignKey,
    Index,
    Integer,
    Numeric,
    String,
    UniqueConstraint,
)
from sqlalchemy.orm import Mapped, mapped_column

from clientbridge.core.db import Base
from clientbridge.models.base import BusinessScoped, PKMixin, TimestampMixin, enum_check


class Invoice(PKMixin, BusinessScoped, TimestampMixin, Base):
    __tablename__ = "invoices"
    __table_args__ = (
        UniqueConstraint("business_id", "number", name="uq_invoices_business_number"),
        enum_check("invoices", "status", "draft", "sent", "partial", "paid", "overdue", "void"),
        Index("ix_invoices_client", "business_id", "client_id"),
        Index("ix_invoices_status", "business_id", "status"),
    )

    client_id: Mapped[str] = mapped_column(ForeignKey("clients.id"), nullable=False)
    number: Mapped[int | None] = mapped_column(BigInteger)
    status: Mapped[str] = mapped_column(String, default="draft", nullable=False)
    currency: Mapped[str] = mapped_column(String(3), default="CAD", nullable=False)
    subtotal_cents: Mapped[int] = mapped_column(BigInteger, default=0, nullable=False)
    tax_total_cents: Mapped[int] = mapped_column(BigInteger, default=0, nullable=False)
    total_cents: Mapped[int] = mapped_column(BigInteger, default=0, nullable=False)
    amount_paid_cents: Mapped[int] = mapped_column(BigInteger, default=0, nullable=False)
    balance_cents: Mapped[int] = mapped_column(BigInteger, default=0, nullable=False)
    issued_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True))
    due_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True))
    paid_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True))
    voided_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True))
    notes: Mapped[str | None] = mapped_column(String)


class Estimate(PKMixin, BusinessScoped, TimestampMixin, Base):
    __tablename__ = "estimates"
    __table_args__ = (
        UniqueConstraint("business_id", "number", name="uq_estimates_business_number"),
        enum_check("estimates", "status", "draft", "sent", "accepted", "declined", "expired"),
        Index("ix_estimates_status", "business_id", "status"),
    )

    client_id: Mapped[str] = mapped_column(ForeignKey("clients.id"), nullable=False)
    number: Mapped[int | None] = mapped_column(BigInteger)
    status: Mapped[str] = mapped_column(String, default="draft", nullable=False)
    subtotal_cents: Mapped[int] = mapped_column(BigInteger, default=0, nullable=False)
    tax_total_cents: Mapped[int] = mapped_column(BigInteger, default=0, nullable=False)
    total_cents: Mapped[int] = mapped_column(BigInteger, default=0, nullable=False)
    valid_until: Mapped[date | None] = mapped_column(Date)
    accepted_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True))
    declined_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True))
    converted_invoice_id: Mapped[str | None] = mapped_column(ForeignKey("invoices.id"))
    notes: Mapped[str | None] = mapped_column(String)


class Line(PKMixin, BusinessScoped, TimestampMixin, Base):
    __tablename__ = "lines"
    __table_args__ = (
        enum_check("lines", "parent_type", "invoice", "estimate"),
        Index("ix_lines_parent", "parent_type", "parent_id"),
    )

    parent_type: Mapped[str] = mapped_column(String, nullable=False)
    parent_id: Mapped[str] = mapped_column(String, nullable=False)
    description: Mapped[str] = mapped_column(String, nullable=False)
    item_id: Mapped[str | None] = mapped_column(ForeignKey("items.id"))
    booking_id: Mapped[str | None] = mapped_column(ForeignKey("bookings.id"))
    quantity: Mapped[float] = mapped_column(Numeric, default=1, nullable=False)
    unit_amount_cents: Mapped[int] = mapped_column(BigInteger, default=0, nullable=False)
    amount_cents: Mapped[int] = mapped_column(BigInteger, default=0, nullable=False)
    tax_rate_id: Mapped[str | None] = mapped_column(ForeignKey("tax_rates.id"))
    tax_amount_cents: Mapped[int] = mapped_column(BigInteger, default=0, nullable=False)
    position: Mapped[int] = mapped_column(Integer, default=0, nullable=False)


class TaxRate(PKMixin, TimestampMixin, Base):
    """System-seeded; business_id NULL = global default."""

    __tablename__ = "tax_rates"
    __table_args__ = (
        enum_check("tax_rates", "jurisdiction", "GST", "HST", "PST", "QST"),
        Index("ix_tax_rates_province", "province", "jurisdiction"),
    )

    business_id: Mapped[str | None] = mapped_column(ForeignKey("businesses.id"), index=True)
    jurisdiction: Mapped[str] = mapped_column(String, nullable=False)
    province: Mapped[str] = mapped_column(String, nullable=False)
    rate_bps: Mapped[int] = mapped_column(Integer, nullable=False)
    name: Mapped[str] = mapped_column(String, nullable=False)

File: frontend/packages/app-core/src/domain/billing.ts
import { useQuery } from "@powersync/react";
import { useEffect, useState } from "react";

import { useAsyncAction } from "../hooks/useAsyncAction";
import type { ApiLike } from "../util/api";
import { blankToNull } from "../util/format";
import type { Intent } from "../util/primitives";
import { strings } from "../strings";

export interface InvoiceRow {
    id: string;
    client_id: string;
    client_name: string | null;
    number: number | null;
    status: string;
    total_cents: number | null;
    balance_cents: number | null;
    issued_at: string | null;
    due_at: string | null;
    pay_token: string | null;
    created_at: string;
}

export interface EstimateRow {
    id: string;
    client_id: string;
    client_name: string | null;
    number: number | null;
    status: string;
    total_cents: number | null;
    valid_until: string | null;
    converted_invoice_id: string | null;
    created_at: string;
}

export interface LineRow {
    id: string;
    description: string;
    quantity: number;
    unit_amount_cents: number;
    amount_cents: number;
    tax_amount_cents: number;
    position: number;
}

const INVOICES_SQL = `
SELECT i.id, i.client_id, c.name AS client_name, i.number, i.status,
       i.total_cents, i.balance_cents, i.issued_at, i.due_at, i.pay_token, i.created_at
FROM invoices i
LEFT JOIN clients c ON c.id = i.client_id
ORDER BY i.created_at DESC`;

const ESTIMATES_SQL = `
SELECT e.id, e.client_id, c.name AS client_name, e.number, e.status,
       e.total_cents, e.valid_until, e.converted_invoice_id, e.created_at
FROM estimates e
LEFT JOIN clients c ON c.id = e.client_id
ORDER BY e.created_at DESC`;

const LINES_SQL = `
SELECT id, description, quantity, unit_amount_cents, amount_cents, tax_amount_cents, position
FROM lines WHERE parent_type = ? AND parent_id = ? ORDER BY position`;

export function useInvoices(): InvoiceRow[] {
    return useQuery<InvoiceRow>(INVOICES_SQL).data;
}

export function useEstimates(): EstimateRow[] {
    return useQuery<EstimateRow>(ESTIMATES_SQL).data;
}

export function useLines(parentType: "invoice" | "estimate", parentId: string): LineRow[] {
    return useQuery<LineRow>(LINES_SQL, [parentType, parentId]).data;
}

export function filterInvoices(rows: InvoiceRow[], q: string): InvoiceRow[] {
    const t = q.trim().toLowerCase();
    if (!t) return rows;
    return rows.filter(
        (r) =>
            (r.client_name ?? "").toLowerCase().includes(t) ||
            (r.number !== null && String(r.number).includes(t)) ||
            r.status.includes(t),
    );
}

export function filterEstimates(rows: EstimateRow[], q: string): EstimateRow[] {
    const t = q.trim().toLowerCase();
    if (!t) return rows;
    return rows.filter(
        (r) =>
            (r.client_name ?? "").toLowerCase().includes(t) ||
            (r.number !== null && String(r.number).includes(t)) ||
            r.status.includes(t),
    );
}

// The status → visual-intent decision is shared; each platform maps the intent to its own tokens.
export function invoiceStatusIntent(status: string): Intent {
    switch (status) {
        case "paid":
            return "success";
        case "sent":
            return "accent";
        case "partial":
            return "warning";
        case "overdue":
            return "danger";
        default:
            return "neutral"; // draft, void
    }
}

export function estimateStatusIntent(status: string): Intent {
    switch (status) {
        case "accepted":
            return "success";
        case "sent":
            return "accent";
        case "declined":
            return "danger";
        default:
            return "neutral"; // draft, expired
    }
}

export interface LineInput {
    description: string;
    quantity: number;
    unit_amount_cents: number;
}

export interface DocResult {
    id: string;
    status: string;
    number: number | null;
}

export function createInvoice(
    api: ApiLike,
    clientId: string,
    lines: LineInput[],
    notes?: string | null,
): Promise<DocResult> {
    return api.post<DocResult>("/v1/invoices", {
        client_id: clientId,
        lines,
        notes: blankToNull(notes),
    });
}

export function updateInvoice(
    api: ApiLike,
    id: string,
    patch: { lines?: LineInput[]; notes?: string | null },
): Promise<DocResult> {
    return api.patch<DocResult>(`/v1/invoices/${id}`, patch);
}

export function sendInvoice(api: ApiLike, id: string): Promise<DocResult> {
    return api.post<DocResult>(`/v1/invoices/${id}/send`, {});
}

export function voidInvoice(api: ApiLike, id: string): Promise<DocResult> {
    return api.post<DocResult>(`/v1/invoices/${id}/void`, {});
}

export function createEstimate(
    api: ApiLike,
    clientId: string,
    lines: LineInput[],
    notes?: string | null,
): Promise<DocResult> {
    return api.post<DocResult>("/v1/estimates", {
        client_id: clientId,
        lines,
        notes: blankToNull(notes),
    });
}

export function updateEstimate(
    api: ApiLike,
    id: string,
    patch: { lines?: LineInput[]; notes?: string | null },
): Promise<DocResult> {
    return api.patch<DocResult>(`/v1/estimates/${id}`, patch);
}

export function sendEstimate(api: ApiLike, id: string): Promise<DocResult> {
    return api.post<DocResult>(`/v1/estimates/${id}/send`, {});
}

export function acceptEstimate(api: ApiLike, id: string): Promise<DocResult> {
    return api.post<DocResult>(`/v1/estimates/${id}/accept`, {});
}

export function declineEstimate(api: ApiLike, id: string): Promise<DocResult> {
    return api.post<DocResult>(`/v1/estimates/${id}/decline`, {});
}

export function convertEstimate(api: ApiLike, id: string): Promise<DocResult> {
    return api.post<DocResult>(`/v1/estimates/${id}/convert`, {});
}

export type DocActionKey = "send" | "void" | "accept" | "decline" | "convert";

export type DocTab = "invoices" | "estimates";

/** Button copy for each document action — shared so web + mobile can't drift (they had). */
export const DOC_ACTION_LABEL: Record<DocActionKey, string> = {
    send: strings.invoices.actionSend,
    void: strings.invoices.actionVoid,
    accept: strings.invoices.actionAccept,
    decline: strings.invoices.actionDecline,
    convert: strings.invoices.actionConvert,
};

export interface DocAction {
    key: DocActionKey;
    run: () => Promise<DocResult>;
}

// The status → available-action state machine is shared; each platform maps the key to a label.
export function invoiceActions(api: ApiLike, row: InvoiceRow): DocAction[] {
    const out: DocAction[] = [];
    if (row.status === "draft") out.push({ key: "send", run: () => sendInvoice(api, row.id) });
    if (row.status !== "void" && row.status !== "paid")
        out.push({ key: "void", run: () => voidInvoice(api, row.id) });
    return out;
}

export function estimateActions(api: ApiLike, row: EstimateRow): DocAction[] {
    const out: DocAction[] = [];
    if (row.status === "draft") out.push({ key: "send", run: () => sendEstimate(api, row.id) });
    if (row.status === "sent") {
        out.push({ key: "accept", run: () => acceptEstimate(api, row.id) });
        out.push({ key: "decline", run: () => declineEstimate(api, row.id) });
    }
    if ((row.status === "sent" || row.status === "accepted") && row.converted_invoice_id === null)
        out.push({ key: "convert", run: () => convertEstimate(api, row.id) });
    return out;
}

export interface DraftLine {
    description: string;
    quantity: string;
    unit: string;
}

export function toLineInputs(drafts: DraftLine[]): LineInput[] {
    return drafts
        .filter((l) => l.description.trim().length > 0)
        .map((l) => ({
            description: l.description.trim(),
            quantity: Number(l.quantity) || 0,
            unit_amount_cents: Math.round((Number(l.unit) || 0) * 100),
        }));
}

export function lineSubtotalCents(lines: LineInput[]): number {
    return lines.reduce((s, l) => s + Math.round(l.quantity * l.unit_amount_cents), 0);
}

export interface TaxRate {
    id: string;
    jurisdiction: string;
    province: string;
    rate_bps: number;
    name: string;
}

/** Province tax rates (REST today; the abstraction is the single place to move to sync later). */
export function useTaxRates(api: ApiLike): TaxRate[] | null {
    const [rates, setRates] = useState<TaxRate[] | null>(null);
    useEffect(() => {
        void api
            .get<TaxRate[]>("/v1/tax-rates")
            .then(setRates)
            .catch(() => {
                setRates([]);
            });
    }, [api]);
    return rates;
}

export interface KeyedLine extends DraftLine {
    key: string;
}

let lineSeq = 0;
const blankLine = (): KeyedLine => ({
    key: `l${(lineSeq += 1)}`,
    description: "",
    quantity: "1",
    unit: "",
});

export interface DocForm {
    clientId: string;
    setClientId: (v: string) => void;
    lines: KeyedLine[];
    setLine: (key: string, patch: Partial<DraftLine>) => void;
    addLine: () => void;
    removeLine: (key: string) => void;
    notes: string;
    setNotes: (v: string) => void;
    subtotalCents: number;
    busy: boolean;
    error: string | null;
    submit: () => void;
}

/** Shared invoice/estimate builder: client + keyed draft lines + notes + validation + submit.
 *  The platform owns only the row/select widgets. */
export function useDocForm(
    api: ApiLike,
    kind: "invoice" | "estimate",
    onCreated: () => void,
): DocForm {
    const [clientId, setClientId] = useState("");
    const [lines, setLines] = useState<KeyedLine[]>([blankLine()]);
    const [notes, setNotes] = useState("");
    const { busy, error, setError, run } = useAsyncAction();

    const setLine = (key: string, patch: Partial<DraftLine>): void => {
        setLines((ls) => ls.map((l) => (l.key === key ? { ...l, ...patch } : l)));
    };
    const addLine = (): void => {
        setLines((ls) => [...ls, blankLine()]);
    };
    const removeLine = (key: string): void => {
        setLines((ls) => (ls.length > 1 ? ls.filter((l) => l.key !== key) : ls));
    };

    const subtotalCents = lineSubtotalCents(toLineInputs(lines));

    const submit = (): void => {
        const payload = toLineInputs(lines);
        if (clientId.length === 0 || payload.length === 0) {
            setError(strings.invoices.incompleteInvoice);
            return;
        }
        void run(
            () =>
                kind === "invoice"
                    ? createInvoice(api, clientId, payload, notes)
                    : createEstimate(api, clientId, payload, notes),
            { onSuccess: onCreated, errorMessage: strings.invoices.saveError },
        );
    };

    return {
        clientId,
        setClientId,
        lines,
        setLine,
        addLine,
        removeLine,
        notes,
        setNotes,
        subtotalCents,
        busy,
        error,
        submit,
    };
}
